docs/design.md

```diff
@@ -1,87 +1,133 @@
+- ebs-csi-controller interacts with K8s storage resources, calls AWS EC2 APIs
+- ebs-csi-node runs on every Node, used by Kubelet to perform privileged operations on attached storage devices
+
+To illustrate these relationships between Kubernetes, the EBS CSI Driver, and AWS, we can look at what happens when you dynamically provision a volume for a given stateful workload:
+
+```mermaid
+sequenceDiagram
+    actor o as Operator
+    participant k as Kubernetes
+    participant c as ebs-csi-controller
+    participant a as AWS EC2 API
+    participant n as ebs-csi-node
+    participant os as Node OS
+
+    o->>k: Create PVC + Pod
+    activate k
+
+    note over k: Pod 'Pending' <br/> PVC 'Pending'
+    k->>c: Notifies about PVC
+    activate c
+    c->>a: Ensure volume created
+    a-->>c: 
+    c-->>k: Create PV handle=<EBS volumeID> <br/> PVC and PV 'bound'
+    deactivate c
+    
+    note over k: Pod 'Scheduled' <br/> VolumeAttachment (VA) created
+    k->>c: Notify about VolumeAttachment (VA)
+    activate c
+    c->>a: Ensure volume attached
+    a-->>c: 
+    c-->>k: Update VA: <br/> Attached=true <br/> DevicePath='/dev/xvdaa'
+    deactivate c
+    
+    note over k: Pod 'ContainerCreating'
+    k->>n: Kubelet Triggers NodeStageVolume
+    activate n
+    opt If Filesystem 
+    n->>os: Format/fsck device
+    os-->>n: 
+    end
+    n->>os: Mount to '/var/lib/kubelet/<pod-id>/volumes/'
+    os-->>n: 
+    n-->>k: Kubelet update Node.volumesAttached
+    deactivate n
+    
+    k->>n: Kubelet Triggers NodePublishVolume
+    activate n
+    n->>os: Bind-mount to container's 'mountPath'
+    os-->>n: 
+    n-->>k: Kubelet update Node.volumesInUse
+    deactivate n
+    
+    note over k: Pod 'Running'
+    k-->>o: Stateful workload running
+    deactivate k
+```
 
-## Requirements
+## CSI Driver Design Requirements
 
 ### Idempotency
 
-All CSI driver calls should be idempotent. A CSI method call with the same parameters must always return the same result. It&#39;s task of CSI driver to ensure that. Examples:
+All CSI calls should be idempotent. The CSI plugin must ensure that a CSI call with the same parameters will always return the same result. Examples:
 
 - CreateVolume call must first check that the requested EBS volume has been already provisioned and return it if so. It should create a new volume only when such volume does not exist.
 - ControllerPublish (=i.e. attach) does not do anything and returns &quot;success&quot; when given volume is already attached to requested node.
 - DeleteVolume does not do anything and returns success when given volume is already deleted (i.e. it does not exist, we don&#39;t need to check that it had existed and someone really deleted it)
 
-Note that it&#39;s task of the CSI driver to make these calls idempotent if related AWS API call is not.
+Note that it&#39;s task of the ebs-plugin to make these calls idempotent even if the related AWS API call is not.
```

Comment:
Drop the first half of this diff, `ebs-plugin` is a confusing term and means the same thing as "CSI driver". We should be using what is essentially an internal term with no meaning, when "CSI Driver" is the official and well understood term.

Reply:
I politely initially disagree with this feedback. I think the phrasing 'ebs-plugin' is clearer because that is what the container is called within our driver. Also the upstream csi driver documentation refers to this component (that gets called by sidecars and interacts with storage provider) as driver-plugin, no?

Calling both the whole deployment, and the container itself the EBS CSI Driver confuses many operators new to our driver, including me when I first started. 

Open to hear more thoughts though.

Reply:
Well, sometimes CSI documentation does use plugin to refer to sidecars and company

"Node Plugin

The node component should be deployed on every node in the cluster through a DaemonSet. It consists of the CSI driver that implements the CSI Node service and the node-driver-registrar sidecar container"

Is there a better term we can use to differentiate?

Reply:
> Calling both the whole deployment, and the container itself the EBS CSI Driver confuses many operators new to our driver, including me when I first started.

I would agree this could be confusing, but it is the former ("the whole deployment"), not the latter ("the container itself") that is inaccurate and causing the confusion. I don't think this document describes the `Deployment` or `DaemonSet` anywhere as "the EBS CSI Driver" with no qualifications, but if it does that language should be altered.

Using the term `ebs-plugin` causes confusion because it is not the standard terminology. It is important to get the terminology right here because the term "CSI Driver" is used this way extensively throughout the community and upstream code/documentation (e.g. https://kubernetes-csi.github.io/docs).

Depending on how the EBS CSI Driver is deployed, it may not even have an `ebs-plugin` container. For example, here is how the semi-popular Gardener tool deploys the driver, they use the name `csi-driver` for the driver container: https://github.com/gardener/gardener-extension-provider-aws/blob/master/charts/internal/shoot-system-components/charts/csi-driver-node/templates/daemonset.yaml#L43

Reply:
Let me include a short glossary/terminology ala https://github.com/awslabs/soci-snapshotter/blob/main/docs/glossary.md in a future PR where we can make this distinction crisp. 

> Using the term ebs-plugin causes confusion because it is not the standard terminology.

On the other hand the term 'ebs-plugin container' here can be helpful because it lets operators know exactly which container to look at for troubleshooting. In the glossary section, we can mention that as of v1.37.0, our csi driver is deployed as the `ebs-plugin` container by default. 

What's confusing to newcomers here is that  the whole package of Deployment + Daemonset is often referred to as the EBS CSI Driver add-on. Perhaps a glossary can remedy this. 

> Depending on how the EBS CSI Driver is deployed, it may not even have an ebs-plugin container. For example Gardener.

Do they even deploy the EBS CSI Driver Controller? I consider installation methods not explicitly approved by this repository as out of scope for this document.